### PR TITLE
Add redux

### PR DESCRIPTION
**Pull request by ebarriosara**
- Add actions, reducer and store.
- Modify container.
- Update readme.md.

### CODE REVIEW

File: 03 Add redux/readme.md [hunk 1]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
-		├── member.ts
-	├── memberApi.ts
-	└── index.ts
+npm install redux react-redux --save
 ```
 
-- First, we add an `api subfolder` inside `src`. Then, inside `src/api` we create a new subfolder named `src/api/model`
-
-- Let's start by defining the data model used in our API. Our data source will be a list of GitHub members. As we know from the previous case, we want to be able to display the id, name and profile image (represented as an Url) inside our members page. Thus, the data that we really care to retrieve should hold 1 number property for the id, and 2 string properties for the user login name and avatar image Url respectively . Consequently, we will create a `member.ts` file inside `src/api/model` with the content below:
+- Let's install types for react-redux as well:
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: Let's install the type definitions for react-redux

File: 03 Add redux/readme.md [hunk 2]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
-
-```javascript
-export const fetchMemberList = () : Promise<MemberEntity[]> => {
-
-  return fetch(membersURL)
-            .then(checkStatus)
-            .then(parseJSON)
-            .then(resolveMembers)
-
+- We will also need the middleware redux-thunk for the asynchronous call to the API.
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: We will also need the middleware redux-thunk to handle the asynchronous call to the  Rest API.

File: 03 Add redux/readme.md [hunk 3]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
+  ```
+
+  - In _fetchMemberList.ts_ we will need to import the action definitions that we created in previous step, the member model that we defined for our API and, finally, the function fetchMemberList from our API:
+
+  ```javascript
+  import { MemberEntity } from '../api/model';
+  import { actionsDefs } from '../const';
+  import { fetchMemberList } from '../api'; 
+  ```
+  - As our action is an asynchronous call to an API, we need two functions: one to start the request to the server and a callback once the request is completed. The first function will be named _fetchMemberListRequestStart_ and it is the one that we will export. This function will call _fetchMemberList()_ from our API, but it returns a promise, which is not a pure function. This means that we cannot return it directly to the reducer. A dispatcher is the solution to this problem, it does the magic that allows to insert the result of actions into the reducers wheel. Precisely, the second function that we need is the one provided to the dispatcher. It returns the result of the action, which is the list of members, once the request to the API is completed:
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: Access to the dispatcher inside of the action (provided by redux-thunk middleware)

File: 03 Add redux/readme.md [hunk 4]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
+  import { actionsDefs } from '../const';
+  
+  export interface MemberListState {
+    memberList : MemberEntity[];
+  }
+  
+  interface BaseAction {
+    type : string;
+    payload : any;
+  }
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: Add here a.comment indicating that BaseAction could be.promoted to a common file and be reused in other reducers

File: 03 Add redux/readme.md [hunk 5]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
+  
+  + const createDefaultState = () => ({
+  +   memberList: [],
+  + });
+  ```
+
+  - The reducer is a function that receives a state and an action and returns a new state. Therefore, in our case, we need to check that the action type is correct and return the new state, which is a new object with the member list received from the API:
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: Indicate here that we create a new obiect (immutability)

File: 03 Add redux/readme.md [hunk 6]
```diff
@@ -19,172 +22,354 @@ Install [Node.js and npm](https://nodejs.org/en/) if they are not already instal
...
+
+  ```diff
+  import { MemberEntity } from '../api/model';
+  import { actionsDefs } from '../const';
+  
+  export interface MemberListState {
+    memberList : MemberEntity[];
+  }
```
Comment (brauliodiez) [inline 03 Add redux/readme.md]: Indicate here that we have created an object just to explain how the spread operator works and help us to keep object immutable, but here we.could just set the state directly to an array of members

**Comment by brauliodiez** [comment]
Can you see the review? Is set as pending

**Comment by ebarriosara** [comment]
Thanks for the review!

I've updated comments as suggested.

**Pull request merged**